File: bindings/tomlplus-dotnet/Tomlplus.cs
// TOML+ .NET bindings — P/Invoke over `tomlplus_ffi`.

using System.Runtime.InteropServices;
using System.Text.Json;

namespace Tomlplus;

internal static partial class Native
{
    private const string LibName = "tomlplus_ffi";

    [LibraryImport(LibName, StringMarshalling = StringMarshalling.Utf8)]
    internal static partial IntPtr tomlplus_parse(string source);

    [LibraryImport(LibName)] internal static partial IntPtr tomlplus_to_json(IntPtr doc);
    [LibraryImport(LibName)] internal static partial IntPtr tomlplus_meta_json(IntPtr doc);
    [LibraryImport(LibName)] internal static partial IntPtr tomlplus_vars_json(IntPtr doc);
    [LibraryImport(LibName)] internal static partial IntPtr tomlplus_validate(IntPtr doc);
    [LibraryImport(LibName)] internal static partial IntPtr tomlplus_dumps(IntPtr doc);
    [LibraryImport(LibName)] internal static partial void   tomlplus_free(IntPtr doc);
    [LibraryImport(LibName)] internal static partial void   tomlplus_free_string(IntPtr s);
    [LibraryImport(LibName)] internal static partial IntPtr tomlplus_last_error();
    [LibraryImport(LibName)] internal static partial IntPtr tomlplus_version();

    internal static string LastError()
        => Marshal.PtrToStringUTF8(tomlplus_last_error()) ?? "unknown FFI error";

    internal static string TakeString(IntPtr p)
    {
        if (p == IntPtr.Zero)
            throw new TomlplusException(LastError());
        try   { return Marshal.PtrToStringUTF8(p) ?? string.Empty; }
        finally { tomlplus_free_string(p); }
    }
}

/// <summary>Thrown for parse failures, validation errors, and FFI errors.</summary>
public sealed class TomlplusException : Exception
{
    public TomlplusException(string message) : base(message) {}
}

/// <summary>A `@name: arg` annotation attached to a key.</summary>
public sealed record Annotation(string Name, object? Arg);

/// <summary>A parsed TOML+ document. Implements <see cref="IDisposable"/>.</summary>
public sealed class Document : IDisposable
{
    private IntPtr _handle;
    private readonly object _gate = new();

    internal Document(IntPtr handle) => _handle = handle;

    internal IntPtr Handle
    {
        get
        {
            if (_handle == IntPtr.Zero)
                throw new ObjectDisposedException(nameof(Document));
            return _handle;
        }
    }

    /// <summary>Whole config tree as a <see cref="JsonElement"/>.</summary>
    public JsonElement Config()
        => JsonDocument.Parse(Native.TakeString(Native.tomlplus_to_json(Handle))).RootElement;

    /// <summary>Resolved <c>[vars]</c> section.</summary>
    public JsonElement Vars()
        => JsonDocument.Parse(Native.TakeString(Native.tomlplus_vars_json(Handle))).RootElement;

    /// <summary>Annotation metadata, keyed by dotted path.</summary>
    public Dictionary<string, List<Annotation>> Meta()
    {
        var s = Native.TakeString(Native.tomlplus_meta_json(Handle));
        using var doc = JsonDocument.Parse(s);
        var result = new Dictionary<string, List<Annotation>>();
        foreach (var prop in doc.RootElement.EnumerateObject())
        {
            var list = new List<Annotation>();
            foreach (var elt in prop.Value.EnumerateArray())
            {
                var name = elt.GetProperty("name").GetString() ?? string.Empty;
                object? arg = elt.GetProperty("arg").ValueKind switch
                {
                    JsonValueKind.String => elt.GetProperty("arg").GetString(),
                    JsonValueKind.Number => elt.GetProperty("arg").GetDouble(),
                    JsonValueKind.Array  => elt.GetProperty("arg")
                                              .EnumerateArray()
                                              .Select(e => e.GetString())
                                              .ToList(),
                    JsonValueKind.True   => true,
                    JsonValueKind.False  => false,
                    _ => null,
                };
                list.Add(new Annotation(name, arg));
            }
            result[prop.Name] = list;
        }
        return result;
    }

    /// <summary>Walk a dotted path; returns <c>null</c> if any segment is missing.</summary>
    public JsonElement? Resolve(string path)
    {
        var node = Config();
        foreach (var part in path.Split('.'))
        {
            if (node.ValueKind != JsonValueKind.Object) return null;
            if (!node.TryGetProperty(part, out var next)) return null;
            node = next;
        }
        return node;
    }

    public bool HasAnnotation(string path, string name)
        => Meta().TryGetValue(path, out var anns)
            && anns.Any(a => a.Name == name);

    /// <summary>Return every validation diagnostic for this document.</summary>
    public List<ValidationError> ValidateAll()
    {
        var s = Native.TakeString(Native.tomlplus_validate(Handle));
        return JsonSerializer.Deserialize<List<ValidationError>>(s, JsonOpts) ?? new();
    }

    private static readonly JsonSerializerOptions JsonOpts =
        new() { PropertyNameCaseInsensitive = true };

    public void Dispose()
    {
        lock (_gate)
        {
            if (_handle != IntPtr.Zero)
            {
                Native.tomlplus_free(_handle);
                _handle = IntPtr.Zero;
            }
        }
    }
}

public sealed class ValidationError
{
    public string  Message  { get; set; } = "";
    public string  Severity { get; set; } = "";
    public Span    Span     { get; set; } = new();
}

public sealed class Span
{
    public int Start { get; set; }
    public int End   { get; set; }
}

/// <summary>Top-level entry points.</summary>
public static class TomlplusApi
{
    public static string Version()
        => Marshal.PtrToStringUTF8(Native.tomlplus_version()) ?? "";

    /// <summary>Parse a TOML+ source string.</summary>
    public static Document Parse(string source)
    {
        var h = Native.tomlplus_parse(source);
        if (h == IntPtr.Zero)
            throw new TomlplusException(Native.LastError());
        return new Document(h);
    }

    public static Document Load(string path) => Parse(File.ReadAllText(path));

    /// <summary>Validate; throws on the first failing constraint.</summary>
    public static void Validate(Document doc)
    {
        foreach (var e in doc.ValidateAll())
            if (e.Severity == "error")
                throw new TomlplusException(e.Message);
    }

    public static string Dumps(Document doc)
        => Native.TakeString(Native.tomlplus_dumps(doc.Handle));
}

File: bindings/tomlplus-java/src/main/java/com/tomlplus/Tomlplus.java
package com.tomlplus;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Pointer;

import java.util.List;
import java.util.Map;

/**
 * TOML+ Java bindings — thin wrapper over the {@code tomlplus_ffi} C library
 * loaded via JNA. The JVM needs the shared library on its native search path
 * ({@code -Djna.library.path=...} or the OS loader).
 */
public final class Tomlplus {
    private Tomlplus() {}

    private interface Lib extends Library {
        Lib INSTANCE = Native.load("tomlplus_ffi", Lib.class);

        Pointer tomlplus_parse(String source);
        Pointer tomlplus_to_json(Pointer doc);
        Pointer tomlplus_meta_json(Pointer doc);
        Pointer tomlplus_vars_json(Pointer doc);
        Pointer tomlplus_validate(Pointer doc);
        Pointer tomlplus_dumps(Pointer doc);
        void    tomlplus_free(Pointer doc);
        void    tomlplus_free_string(Pointer s);
        String  tomlplus_last_error();
        String  tomlplus_version();
    }

    private static final ObjectMapper JSON = new ObjectMapper();

    public static String version() {
        return Lib.INSTANCE.tomlplus_version();
    }

    /** Parse a TOML+ source string. */
    public static Document parse(String source) {
        Pointer h = Lib.INSTANCE.tomlplus_parse(source);
        if (h == null) {
            throw new TomlplusException(Lib.INSTANCE.tomlplus_last_error());
        }
        return new Document(h);
    }

    /** Validate and throw on the first failing constraint. */
    public static void validate(Document doc) {
        for (Map<String, Object> err : doc.validateAll()) {
            if ("error".equals(err.get("severity"))) {
                throw new TomlplusException((String) err.get("message"));
            }
        }
    }

    /** Re-serialise to TOML+ text. */
    public static String dumps(Document doc) {
        Pointer p = Lib.INSTANCE.tomlplus_dumps(doc.handle());
        return takeString(p);
    }

    // ── Parsed document handle ───────────────────────────────────────────────

    public static final class Document implements AutoCloseable {
        private Pointer handle;
        private final Object lock = new Object();

        Document(Pointer h) { this.handle = h; }

        Pointer handle() {
            if (handle == null) throw new IllegalStateException("Document is closed");
            return handle;
        }

        /** Whole config tree as a {@code Map<String, Object>}. */
        public Map<String, Object> config() {
            return readJsonMap(Lib.INSTANCE.tomlplus_to_json(handle()));
        }

        /** Resolved {@code [vars]} entries. */
        public Map<String, Object> vars() {
            return readJsonMap(Lib.INSTANCE.tomlplus_vars_json(handle()));
        }

        /** Annotation metadata, keyed by dotted path. */
        @SuppressWarnings("unchecked")
        public Map<String, List<Annotation>> meta() {
            String json = takeString(Lib.INSTANCE.tomlplus_meta_json(handle()));
            try {
                Map<String, List<Map<String, Object>>> raw = JSON.readValue(json, Map.class);
                java.util.LinkedHashMap<String, List<Annotation>> out = new java.util.LinkedHashMap<>();
                for (var e : raw.entrySet()) {
                    var anns = new java.util.ArrayList<Annotation>(e.getValue().size());
                    for (var a : e.getValue()) {
                        anns.add(new Annotation((String) a.get("name"), a.get("arg")));
                    }
                    out.put(e.getKey(), anns);
                }
                return out;
            } catch (Exception ex) {
                throw new TomlplusException("meta JSON parse failed: " + ex.getMessage());
            }
        }

        /** Walk a dotted path; returns {@code null} if any segment is missing. */
        public Object resolve(String path) {
            Object node = config();
            for (String part : path.split("\\.")) {
                if (!(node instanceof Map)) return null;
                node = ((Map<?, ?>) node).get(part);
                if (node == null) return null;
            }
            return node;
        }

        public boolean hasAnnotation(String path, String name) {
            var anns = meta().get(path);
            if (anns == null) return false;
            return anns.stream().anyMatch(a -> name.equals(a.name()));
        }

        @SuppressWarnings("unchecked")
        public List<Map<String, Object>> validateAll() {
            String json = takeString(Lib.INSTANCE.tomlplus_validate(handle()));
            try {
                return JSON.readValue(json, List.class);
            } catch (Exception ex) {
                throw new TomlplusException("validate JSON parse failed: " + ex.getMessage());
            }
        }

        @Override
        public void close() {
            synchronized (lock) {
                if (handle != null) {
                    Lib.INSTANCE.tomlplus_free(handle);
                    handle = null;
                }
            }
        }
    }

    /** A `@name: arg` annotation. */
    public record Annotation(String name, Object arg) {}

    /** Thrown for parse failures, validation errors, and FFI errors. */
    public static final class TomlplusException extends RuntimeException {
        public TomlplusException(String msg) { super(msg); }
    }

    // ── Helpers ─────────────────────────────────────────────────────────────

    @SuppressWarnings("unchecked")
    private static Map<String, Object> readJsonMap(Pointer p) {
        String s = takeString(p);
        try {
            return JSON.readValue(s, Map.class);
        } catch (Exception e) {
            throw new TomlplusException("JSON parse failed: " + e.getMessage());
        }
    }

    private static String takeString(Pointer p) {
        if (p == null) {
            throw new TomlplusException(Lib.INSTANCE.tomlplus_last_error());
        }
        try {
            return p.getString(0, "UTF-8");
        } finally {
            Lib.INSTANCE.tomlplus_free_string(p);
        }
    }
}
